File: Leetcode/19_reverse_linked_list2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function (head, m, n) {
  // 예외 처리
  if (!head || m === n) return head;

  let root = new ListNode();
  let start = root;
  let end = null;
  root.next = head;

  for (let i = 0; i < m - 1; i++) {
    start = start.next;
  }
  end = start.next;

  for (let i = 0; i < n - m; i++) {
    let tmp = start.next;
    start.next = end.next;
    end.next = end.next.next;
    start.next.next = tmp;
  }

  return root.next;
};

File: Leetcode/19_reverse_linked_list2.py
# https://leetcode.com/problems/reverse-linked-list-ii/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reverseBetween(head: ListNode, m: int, n: int) -> ListNode:
    # 예외 처리
    if not head or m == n:
        return head

    root = start = ListNode(None)
    root.next = head

    # 위치를 바꿀 시작 지점까지 포인터 이동하기
    for _ in range(m - 1):
        start = start.next
    end = start.next

    # 종료 위치까지 노드 스위칭 하기
    for _ in range(n - m):
        tmp, start.next, end.next = start.next, end.next, end.next.next
        start.next.next = tmp

    return root.next

